netty/src/main/java/io/grpc/netty/NettyServer.java

```diff
@@ -291,4 +318,48 @@ public ReferenceCounted touch(Object hint) {
       return this;
     }
   }
+
+  /**
+   * A class that can answer channelz queries about the server listen sockets.
+   */
+  private static final class ListenSocket implements Instrumented<SocketStats> {
+    private final LogId id = LogId.allocate(getClass().getName());
+    private final Channel ch;
+
+    ListenSocket(Channel ch) {
+      this.ch = ch;
+    }
+
+    @Override
+    public ListenableFuture<SocketStats> getStats() {
+      final SettableFuture<SocketStats> ret = SettableFuture.create();
+      if (ch.eventLoop().inEventLoop()) {
+        // This is necessary, otherwise we will block forever if we get the future from inside
+        // the event loop.
+        ret.set(new SocketStats(
+            /*data=*/ null,
+            ch.localAddress(),
+            /*remoteAddress=*/ null,
+            /*security=*/ null));
+        return ret;
+      }
+      ch.eventLoop().submit(
```

Comment:
nit: on failure you need to fail `ret`.    In this case, probably adding a listener to the future returned by submit.

Reply:
fixed here and other classes